leakpro/attacks/gia_attacks/huang.py

```diff
@@ -0,0 +1,222 @@
+
+        # Stop updating running statistics
+        for module in self.model.modules():
+            if isinstance(module, torch.nn.BatchNorm2d):
+                module.momentum = 0
+
+        self.reconstruction, self.reconstruction_loader = get_at_images(self.client_loader)
+        self.reconstruction.requires_grad = True
+        client_gradient = self.train_fn(self.model, self.client_loader, self.optimizer, self.criterion, self.epochs)
+        self.client_gradient = [p.detach() for p in client_gradient]
+
+
+    def run_attack(self:Self) -> Union[GIAResults, Generator[tuple[int, Tensor]]]:
+        """Run the attack and return the combined metric result.
+
+        Returns
+        -------
+            GIAResults: Container for results on GIA attacks.
+
+        """
+        optimizer = torch.optim.Adam([self.reconstruction], lr=self.attack_lr)
+        scheduler = torch.optim.lr_scheduler.MultiStepLR(optimizer,
+                                                            milestones=[self.iterations // 2.667, self.iterations // 1.6,
+
+                                                                        self.iterations // 1.142], gamma=0.1)
+        for i in range(self.iterations):
+            # loss function which does training and compares distance from reconstruction training to the real training.
+            closure = self.gradient_closure(optimizer)
+
+            loss = optimizer.step(closure)
+            scheduler.step()
+            with torch.no_grad():
+                self.reconstruction.data = torch.max(
+                    torch.min(self.reconstruction, (1 - self.data_mean) / self.data_std), -self.data_mean / self.data_std
+                    )
+                if (i +1) % 500 == 0 and self.median_pooling:
+                    self.reconstruction.data = MedianPool2d(kernel_size=3, stride=1, padding=1, same=False)(self.reconstruction)
+            if i % 250 == 0:
+                logger.info(f"Iteration {i}, loss {loss}")
+                yield i, dataloaders_ssim_ignite(self.client_loader, self.reconstruction_loader)
+            # Chose image who has given least loss
+            if loss < self.best_loss:
+                self.best_loss = loss
+                self.best_reconstruction = deepcopy(self.reconstruction_loader)
+                self.best_reconstruction_round = i
+                logger.info(f"New best loss: {loss} on round: {i}")
+
+        return GIAResults(self.client_loader, self.best_reconstruction,
+                          ssim_score=dataloaders_ssim_ignite(self.client_loader, self.reconstruction_loader),
+                          data_mean=self.data_mean, data_std=self.data_std)
+
+
+    def gradient_closure(self: Self, optimizer: torch.optim.Optimizer) -> Callable:
+        """Returns a closure function that performs a gradient descent step.
+
+        The closure function computes the gradients, calculates the reconstruction loss,
```

Comment:
Update this comment, it is copied from the other attack. Explain the loss_r_feature.